setup.sh

```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+# Ensure script is run as root
+if [ "$(id -u)" != "0" ]; then
+   echo "This script must be run as root (use sudo)." 
+   exit 1
+fi
+
+# Install necessary packages based on the package manager available
+if [ -x "$(command -v apt-get)" ]; then
+    echo "Detected Debian-based system. Installing dependencies using apt-get."
+    sudo apt-get update
+    sudo apt-get install -y git python3-pip llvm cmake curl
+
+elif [ -x "$(command -v yum)" ]; then
+    echo "Detected Red Hat-based system. Installing dependencies using yum."
+    sudo yum install -y git python3-pip llvm cmake curl
+
+elif [ -x "$(command -v pacman)" ]; then
+    echo "Detected Arch-based system. Installing dependencies using pacman."
+    sudo pacman -Sy --needed git python-pip llvm cmake curl
+
+elif [ -x "$(command -v brew)" ]; then
+    echo "Detected macOS. Installing dependencies using Homebrew."
+    brew install git python3 llvm cmake curl
+
+else
+    echo "Unsupported OS or package manager. Please install dependencies manually."
+    exit 1
+fi
```

Comment:
Stacking if-elif-elif-...-else unnecesarily increase the maintanance burden. On Linux it's standard to use Linux Standard base (lsb) to identify the linux distribution and release as e.g. there are scenarios where pacman is used on multiple distributions that might not have the same packages or their naming e.g. GNU Parabola

Reply:
Both bash and zsh have `OSTYPE` environment variable, which can be used along the lines of:

```bash
# Grep for 'darwin':
if [[ $OSTYPE =~ "darwin" ]]; then
    echo "macOS"
else
    echo "not macOS"
fi
```

That said, upstream projects never should modify system state as part of any script. At most it could suggest packages but not perform any action. It's a security border IMHO that should not be crossed.
 